cpp/src/arrow/python/arrow_to_python.cc

```diff
@@ -0,0 +1,172 @@
+    case Type::STRUCT: {
+      auto s = std::static_pointer_cast<StructArray>(arr);
+      auto l = std::static_pointer_cast<ListArray>(s->field(0));
+      if (s->type()->child(0)->name() == "list") {
+        return DeserializeList(l->values(), l->value_offset(index),
+            l->value_offset(index + 1), base, tensors, result);
+      } else if (s->type()->child(0)->name() == "tuple") {
+        return DeserializeTuple(l->values(), l->value_offset(index),
+            l->value_offset(index + 1), base, tensors, result);
+      } else if (s->type()->child(0)->name() == "dict") {
+        return DeserializeDict(l->values(), l->value_offset(index),
+            l->value_offset(index + 1), base, tensors, result);
+      } else {
+        DCHECK(false) << "error";
+      }
+    }
+    // We use an Int32Builder here to distinguish the tensor indices from
+    // the Type::INT64 above (see tensor_indices_ in sequence.h).
+    case Type::INT32: {
+      return DeserializeArray(arr, index, base, tensors, result);
+    }
+    default:
+      DCHECK(false) << "union tag not recognized " << type;
+  }
+  return Status::OK();
+}
+
+#define DESERIALIZE_SEQUENCE(CREATE, SET_ITEM)                              \
+  auto data = std::dynamic_pointer_cast<UnionArray>(array);                 \
+  int32_t size = array->length();                                           \
+  PyObject* result = CREATE(stop_idx - start_idx);                          \
+  auto types = std::make_shared<Int8Array>(size, data->type_ids());         \
+  auto offsets = std::make_shared<Int32Array>(size, data->value_offsets()); \
+  for (int32_t i = start_idx; i < stop_idx; ++i) {                          \
+    if (data->IsNull(i)) {                                                  \
+      Py_INCREF(Py_None);                                                   \
+      SET_ITEM(result, i - start_idx, Py_None);                             \
+    } else {                                                                \
+      int32_t offset = offsets->Value(i);                                   \
+      int8_t type = types->Value(i);                                        \
+      std::shared_ptr<Array> arr = data->child(type);                       \
+      PyObject* value;                                                      \
+      RETURN_NOT_OK(get_value(arr, offset, type, base, tensors, &value));   \
+      SET_ITEM(result, i - start_idx, value);                               \
+    }                                                                       \
+  }                                                                         \
+  *out = result;                                                            \
+  return Status::OK();
+
+Status DeserializeList(std::shared_ptr<Array> array, int32_t start_idx, int32_t stop_idx,
+    PyObject* base, const std::vector<std::shared_ptr<Tensor>>& tensors, PyObject** out) {
+  DESERIALIZE_SEQUENCE(PyList_New, PyList_SetItem)
```

Comment:
Ok, I made it a template now and then I can't do PyList_SET_ITEM any more since it is a macro; I'd suggest we keep it a template for now and if this will be a performance bottleneck later, we make it a macro again?